### PR TITLE
Use local info of neighbors, Global Camera, Simulate gravity for obstacles

### PR DESCRIPTION
**Pull request by Zhehui-Huang**

### CODE REVIEW

**Review by SumeetBatra — APPROVED** [review]
I see no glaring issues with the "local info of neighbors" code. Good stuff! 
I will refer to @alex-petrenko  for the camera and obstacles code as I'm not super familiar with those bits.

File: gym_art/quadrotor_multi/quadrotor_multi.py [hunk 1]
```diff
@@ -20,18 +20,23 @@ def __init__(self,
...
+                 adaptive_env=False, obstacle_traj='gravity', neighbor_info_mode='local', num_use_neighbor_obs=0):
 
         super().__init__()
 
         self.num_agents = num_agents
         self.swarm_obs = swarm_obs
+        if neighbor_info_mode == "global":
+            self.num_use_neighbor_obs = self.num_agents - 1
+        else:
+            self.num_use_neighbor_obs = num_use_neighbor_obs
```
Comment (alex-petrenko) [inline gym_art/quadrotor_multi/quadrotor_multi.py:39]: This feels like an extra parameter that we don't really need.
I think num_use_neighbor_obs should default to -1, which means all agents, and value = 0 will mean blind agents.
Although it is a relatively minor thing, I think we can leave it like this for now

File: gym_art/quadrotor_multi/quadrotor_multi.py [hunk 2]
```diff
@@ -165,6 +185,24 @@ def extend_obs_space(self, obs):
...
+            rel_pos = np.linalg.norm(obs_neighbor_rel[:, :3], axis=1)
+            rel_pos_index = rel_pos.argsort()
+            obs_neighbor_rel_n_close = np.array(
+                [obs_neighbor_rel[rel_pos_index[i]] for i in range(self.num_use_neighbor_obs)])
+            obs_neighbors.append(obs_neighbor_rel_n_close.reshape(-1))
+        obs_neighbors = np.stack(obs_neighbors)
+        # clip observation space of neighborhoods
+        obs_neighbors = np.clip(
+            obs_neighbors, a_min=self.clip_neighbor_space_min_box, a_max=self.clip_neighbor_space_max_box,
+        )
```
Comment (alex-petrenko) [inline gym_art/quadrotor_multi/quadrotor_multi.py:202]: Why do we clip it here but not in extend_obs_space?

File: gym_art/quadrotor_multi/quadrotor_multi.py [hunk 3]
```diff
@@ -205,7 +243,10 @@ def reset(self):
 
         # extend obs to see neighbors
         if self.swarm_obs and self.num_agents > 1:
-            obs_ext = self.extend_obs_space(obs)
+            if self.num_use_neighbor_obs == (self.num_agents - 1):
+                obs_ext = self.extend_obs_space(obs)
+            else:
+                obs_ext = self.extend_obs_space_n_closest(obs)
```
Comment (alex-petrenko) [inline gym_art/quadrotor_multi/quadrotor_multi.py:249]: This is not 100% correct (I think).
When we're going to use attention with positional encoding, it is going to be important to sort the drones by distance Even when num_use_neighbor_obs == (self.num_agents - 1) (to provide proper positional values).

This is important. I will accept this by now, but it should be taken into account when we introduce attention.

File: gym_art/quadrotor_multi/quadrotor_multi.py [hunk 4]
```diff
@@ -237,7 +279,10 @@ def step(self, actions):
             self.pos[i, :] = self.envs[i].dynamics.pos
 
         if self.swarm_obs and self.num_agents > 1:
-            obs_ext = self.extend_obs_space(obs)
+            if self.num_use_neighbor_obs == (self.num_agents - 1):
+                obs_ext = self.extend_obs_space(obs)
+            else:
+                obs_ext = self.extend_obs_space_n_closest(obs)
```
Comment (alex-petrenko) [inline gym_art/quadrotor_multi/quadrotor_multi.py:285]: Same here. See line 246

**Review by alex-petrenko — APPROVED** [review]
I haven't thoroughly reviewed Camera or Obstacles code. I trust you that it works, and regular envs without obstacles are more crucial now.

When we start training envs with obstacles, we should do another pass together on all the code that involves obstacles

File: gym_art/quadrotor_multi/quadrotor_multi.py (hunk 2 above)
Comment (Zhehui-Huang) [reply gym_art/quadrotor_multi/quadrotor_multi.py:202]: Actually, we do clip the obs_neighbors in extend_obs_space. But here, this function is for using the local info of neighbor quadrotors. And we also have to clip the obs_neighbors here.

File: gym_art/quadrotor_multi/quadrotor_multi.py (hunk 3 above)
Comment (Zhehui-Huang) [reply gym_art/quadrotor_multi/quadrotor_multi.py:249]: Sure. I will take care of this.

**Pull request merged**